### PR TITLE
Move default template to static `pyproject.toml`, take 2

### PR DESCRIPTION
**Pull request by astrojuanlu**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Alternative to gh-2569, exploring a more careful approach.

Beginning of gh-2280. Note that this should be complemented by a similar PR in https://github.com/kedro-org/kedro-starters/, and is conditioned by the outcomes of gh-2388. But at least it shows how it would be done. Happy to postpone this PR to a later stage if needed.

## Development notes

In order, the commits do the following:

1. Cosmetic changes: remove `environment.yml` mention in the template README.md, rename `setup.cfg` to `.flake8` to reduce confusion, replace `pip install .[pandas.CSVDataSet]` with `pip install kedro-datasets[pandas.CSVDataSet]` (technically out of scope, but harmless).
2. Replace `src/setup.py` with `src/pyproject.toml`. Technically this would be the end, but at this point there are 2 `pyproject.toml` files in the template, so more work is done to reduce confusion.
3. ~~Attempt the `requirements.txt` changes as we decided in https://github.com/kedro-org/kedro/pull/2569#issuecomment-1561189348~~ (skipping for now, and trying to focus on changing the layout without changing the effect of `pip install -r src/requirements.txt`).
4. ~~Merge the two `pyproject.toml` (essentially move one of them one directory up) to have only one project metadata file and turn the template into a Pythonic src-layout package~~ (edit: this will be done in another PR, see discussion)

There are no new tests for this change, the goal is that existing ones pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -226,6 +226,8 @@
     "https://opensource.org/license/apache2-0-php/",
     "https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password",
     "https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/api/snowflake.snowpark.DataFrameWriter.saveAsTable.html"
+    # temporarily, until the file actually exists in `main`
+    "https://github.com/kedro-org/kedro/blob/main/kedro/templates/project/%7B%7B%20cookiecutter.repo_name%20%7D%7D/.flake8",
```
Comment (astrojuanlu) [inline docs/source/conf.py:230]: Apparently this has no effect in preventing the warning (then turned into an error) 🤔 Will ignore it for now

**Comment by merelcht** [comment]
To comment on the description and specifically the breaking changes from moving `tests` and `pyproject.toml`: I think it would be best to leave that last step and do it as part of `0.19.0`. We're removing some of the commands affected in `0.19.0` and I feel that making it all backwards compatible is probably a bit of a wasted effort.

**Comment by astrojuanlu** [comment]
As per the latest discussion, I have removed the breaking changes from this branch and kept only the non-breaking ones.

This should be ready for review now.

**Comment by astrojuanlu** [comment]
e2e tests failing because of a release with https://github.com/kedro-org/kedro-plugins/pull/302 is needed

File: features/steps/test_starter/{{ cookiecutter.repo_name }}/src/requirements.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@ isort~=5.0
 jupyter~=1.0
 jupyterlab_server>=2.11.1, <2.16.0
 jupyterlab~=3.0, <3.6.0
-kedro[pandas.CSVDataSet]=={{ cookiecutter.kedro_version }}
+kedro-datasets[pandas.CSVDataSet]~=1.2
```
Comment (merelcht) [inline features/steps/test_starter/{{ cookiecutter.repo_name }}/src/requirements.txt:9]: This isn't actually related to the move to `pyproject.toml` right? It's not a big issue to add it here, but in general I think we shouldn't be mixing changes in PRs that aren't actually related.

**Review by merelcht — APPROVED** [review]
LGTM! 👍 This change should also be added in the release notes.

File: features/steps/test_starter/{{ cookiecutter.repo_name }}/src/requirements.txt (hunk 2 above)
Comment (astrojuanlu) [reply features/steps/test_starter/{{ cookiecutter.repo_name }}/src/requirements.txt:9]: Fair enough, I'm going to drop this commit and open an issue about it

Comment (astrojuanlu) [reply features/steps/test_starter/{{ cookiecutter.repo_name }}/src/requirements.txt:9]: https://github.com/kedro-org/kedro/issues/2933

File: docs/source/get_started/kedro_concepts.md [hunk 3]
```diff
@@ -72,7 +72,7 @@ project-dir         # Parent directory of the template
 ├── notebooks       # Project-related Jupyter notebooks (can be used for experimental code before moving the code to src)
 ├── pyproject.toml  # Identifies the project root and contains configuration information
 ├── README.md       # Project README
-├── setup.cfg       # Configuration options for `pytest` when doing `kedro test` and for the `isort` utility when doing `kedro lint`
+├── .flake8         # Configuration options for `flake8` (linting)
```
[the remark this answers is not in the document]
Comment (astrojuanlu) [reply docs/source/get_started/kedro_concepts.md:75]: Hey @DimedS , thanks for the review! Good catch, I made the change for the `test_starter` but not for the template itself. Will push it now.

[the remark this answers is not in the document]
Comment (astrojuanlu) [reply docs/source/get_started/kedro_concepts.md:75]: Done

**Review by DimedS — APPROVED** [review]
Thank you, @astrojuanlu . Everything looks good. I'll apply that approach to all the starters.

**Comment by astrojuanlu** [comment]
Let's hold this one after 0.18.13 is released 👍🏽 so we have time to coordinate with the starters then

**Pull request merged**